src/delta/delta.c

```diff
@@ -328,47 +324,94 @@ static int enumerate_dir(Hashmap *top, Hashmap *bottom, Hashmap *drops, const ch
 
+                p = strjoin(path, "/", de->d_name, NULL);
+                if (!p)
+                        return -ENOMEM;
+
                 if (dropins && de->d_type == DT_DIR && endswith(de->d_name, ".d"))
-                        enumerate_dir_d(top, bottom, drops, path, de->d_name);
+                        enumerate_dir_d(bottom, p, de->d_name);
 
                 if (!dirent_is_file(de))
                         continue;
 
-                p = strjoin(path, "/", de->d_name, NULL);
-                if (!p)
-                        return -ENOMEM;
-
-                log_debug("Adding at top: %s %s %s", basename(p), draw_special_char(DRAW_ARROW), p);
-                k = hashmap_put(top, basename(p), p);
-                if (k >= 0) {
-                        p = strdup(p);
-                        if (!p)
-                                return -ENOMEM;
-                } else if (k != -EEXIST) {
-                        free(p);
+                k = unit_name_template(basename(p), &tmp);
+                if (k < 0 && k != -EINVAL)
                         return k;
-                }
 
-                log_debug("Adding at bottom: %s %s %s", basename(p), draw_special_char(DRAW_ARROW), p);
-                free(hashmap_remove(bottom, basename(p)));
-                k = hashmap_put(bottom, basename(p), p);
+                key = tmp ?: basename(p);
+
+                log_debug("Adding: top[%s] %s %s", key, draw_special_char(DRAW_ARROW), p);
+                k = hashmap_put(top, key, p);
+
                 if (k < 0) {
-                        free(p);
-                        return k;
+                        if (k == -EEXIST) {
+                                p = strdup(p);
+                                if (!p)
+                                        return -ENOMEM;
+                                if (tmp) {
+                                        key = strdup(tmp);
+                                        if (!key)
+                                                return -ENOMEM;
+                                } else
+                                        key = basename(p);
+
+                                drops = hashmap_get(bottom, key);
+                                if (!drops) {
+                                        drops = hashmap_new(&string_hash_ops);
+                                        if (!drops)
+                                                return -ENOMEM;
+
+                                        log_debug("Adding: bottom[%s] %s drops[%s] %s []",
+                                                  key, draw_special_char(DRAW_ARROW),
+                                                  p, draw_special_char(DRAW_ARROW));
+
+                                        k = hashmap_put(drops, p, NULL);
+                                        if (k < 0)
+                                                return k;
+
+                                        key = strdup(key);
```

Comment:
Same here. `key` holds what was returned by `strdup(key);` above, and the pointer to this is now leaked. Try running valgrind on that tool.

I think you need to refactor that code a bit so the life time of the allocated strings become more obvious.